### DISCUSSION TITLE
Deleting disputed territories in a nominatim instance

### DISCUSSION THREAD

**Discussion by kenseii** [osm-search/Nominatim#2389] [Q&A] [answered]
Hello

Is there a way of deleting pois in disputed territories(for example pois in islands) in nominatim instance having their polygons? I am thinking of deleting from place and placex tables but I am not sure it will be enough to not display those info.

I know it might not be a good idea to delete the data but as nominatim doesn't have a standard way of dealing with them i think it is the only solution.

**Comment by lonvia** [comment, marked as answer]
There is an undocumented maintenance SQL function `place_force_delete()` which you can use to remove a relation properly. Find the place_id of the relation in question, for example, by using the `/details` endpoint. Run `SELECT place_force_delete(<place_id>)` in psql and then reindex your database with `nominatim index` to make sure all places that are affected (because they have the unwanted place in their address) are updated as well.

The big disadvantage of this approach is that it undoes the previous indexing and that is expensive and doesn't always work perfectly. Also, if you run updates, then the relation might just return the moment somebody edits it.

The better approach would be to filter the relations before they go into the database. There is a modestly simple way to achieve that with [osmium-tool](https://osmcode.org/osmium-tool/). Create a change file in OPL format that deletes all the relations/ways you don't want to have in the database. This is what it would look like:

```
w99563 v9999 dD
r18432 v9999 dD
r185736 v9999 dD
```

The `v9999` ensures that the object in your changefile always have a higher version than whatever is in your planet file and thus take precedence. Make sure the objects are ordered: nodes, ways, then relations. Each group ordered by ID.

Then apply this change file to your planet:

```
osmium apply-changes planet.osm.pbf blacklist.opl -o planet-filtered.osm.pbf
```

Now import the filtered planet.

The same works for change files. Download them separately, for example, using [pyosmium](https://docs.osmcode.org/pyosmium/latest/updating_osm_data.html). Then apply your blacklist:

```
osmium merge-changes -s changefile.osc.gz blacklist.opl -o changefile-filtered.osc.gz
```

And that you can apply manually to Nominatim:

```
nominatim add-data --diff changefile-filtered.osc.gz
nominatim index
```

Nominatim really should get native support for this kind of blacklisting but that has to wait until we have #2047.

**Reply by kenseii** [reply]
@lonvia Thank you for the guidance.
As i already have the data imported in the db i think i will start with the `place_force_delete()` and reindexing option.

About the osmium option, is it possible to use it with nominatim continuous updates without having to reimport planet data again?

**Reply by lonvia** [reply]
Yes, you can delete the objects and then switch your update process to use the filtered changefile in the future.

**Reply by kenseii** [reply]
Thank you. Which file do i have to change for the update process to use the filtered changefile?

**Reply by lonvia** [reply]
You have to write your own update script, I'm afraid. Something along the lines of https://github.com/osm-search/Nominatim/blob/master/utils/update_database.sh but with the commands outlined above.

**Reply by kenseii** [reply]
Thanks i will give it a try. by the way are all the pois in both placex and place? because i deleted the ones in placex but it seems some data are still existing.

**Reply by kenseii** [reply]
My bad, it seems that the data were correctly deleted. Thank you very much.
